### PR TITLE
IBL shadowing

### PR DESCRIPTION
**Pull request by MiiBond**
This PR adds an `IblShadowsRenderPipeline` that adds shadowing for image-based lighting (IBL). It's applied as a post process after rendering your scene.

All control over the shadows will be done through the `IblShadowsRenderPipeline` object. After creating the pipeline, the first things that the user needs to do are:

1. Assign an IBL image with setIblTexture . This function can take either a cube map or an equirectangular map.
2. Add any meshes that you don't want casting shadows with addExcludedMesh . The apparent resolution of the shadows is directly tied to the size of the scene so you want to only include meshes that are important for shadows. e.g. a skybox should be excluded to avoid making the shadow-casting world huge as well as casting a shadow over everything.
3. Update the scene bounds whenever a shadow-casting object moves using updateSceneBounds. This will recalculate the size of the scene and trigger voxelization. This is a pretty expensive operation in WebGL 2.0 so animated shadow-casters are probably not practical.

There are several pieces to how the effect is created. First, there are two passes that are only run when needed:

* `_IblShadowsVoxelRenderer` - Responsible for building a voxel grid of the geometry in the scene as a 3D texture. This texture is used when tracing rays to the IBL and determining what is in shadow and what isn't. The voxelization only needs to be done when some geometry in the scene moves or is added/removed.
* `_IblShadowsImportanceSamplingRenderer` - Responsible for building the CDF (cumulative distribution function) maps that will make it easy to pick the most important ray directions while sampling. This is only run when the IBL changes.

The `IblShadowsRenderPipeline` also requires several maps to be output from the prepass renderer, including world normals, world positions, velocity and depth. We should change this logic to render these per camera because, right now, the prepass renderer only supports a single camera.

The `IblShadowsRenderPipeline` creates four post process passes that do the per-frame shadow calculations and apply them to the scene:

1. `_IblShadowsVoxelTracingPass` - this full-screen pass traces rays through the voxel grid using several of the prepass targets and the IBL CDF. In addition to voxel tracing, it also calculates screen-space shadows to catch small details and combines this with the lower-res shadows from the voxel tracing. This pass outputs a pretty noisy, greyscale approximation of the scene's shadows.
2. `_IblShadowsSpatialBlurPass` - this full-screen pass blurs the shadow texture.
3. `_IblShadowsAccumulationPass` - this full-screen pass accumulates the shadows over multiple frames. It copies both the accumulation and position buffers from the previous frame and uses them (in conjunction with the velocity buffer) to do this.

The output of the accumulation pass is composited onto the scene with another post-process effect.

Here's a PG to test with:
https://playground.babylonjs.com/#8R5SSE#197

**Some important exposed settings:**
* voxel resolution - the dimension of the voxel grid, in voxels. This directly affects how sharp the shadows are. Must be power-of-two.
* sample directions - how many samples of the voxel grid per pixel. Increasing this results in less noise but adds significantly to the per-frame cost of shadows
* toggling screen-space shadows - SS shadows add small shadow detail close to the shadow caster but may not be needed in all scenes so this is togglable and has various settings to control how they behave.

**Updating the voxel grid:**
* When a mesh is added or removed from a scene, the voxelization of the scene is automatically rebuilt.
* When a mesh moves in the scene, you must call `updateSceneBounds` to get the shadows to rebuild. Is there some place that I could do this automatically for users?

**Things that still need work:**
* The shadows are greyscale and are an approximation of the light intensity blocked by an object. They often over-darken surfaces because of this. I have plans to improve this and also add better support for coloured lighting.
* IBL shadows don't currently work with multiple cameras (mostly due to this issue https://forum.babylonjs.com/t/viewport-issue-when-using-prepassrenderer/51276)
* WebGPU is not yet fully supported. When WebGPU is working, I should be able to make use of compute shaders to significantly speed up voxel grid generation. This should make dynamic scene much more practical.
* Various performance and quality tweaks. e.g. support downsizing the buffers for faster speed.
IBL Shadows

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15106/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15106/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15106/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15106/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15106/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15106/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15106/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15106/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15106/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15106/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15106/merge/testResults/webgpuplaywright/index.html

**Comment by MiiBond** [comment]
Would it be possible to get some general feedback about whether this is the right approach for implementing something like this in Babylon? I've modeled this after the DepthPeelingRenderer, which is a scene component, but maybe this should be a pipeline?
The effect has many passes and parts so I'm curious whether I'm structuring them in a reasonable way inside Babylon.

**Comment by Popov72** [comment]
I'll take a look at it early next week.

**Comment by Popov72** [comment]
Here are some quick comments, but not a full review, as I probably won't be able to follow up because I'll be away for a few months in a couple of days, and it's only in draft mode:
* "clip" in some variable / define names (like `viewDepthClipSpace` or `PREPASS_CLIPSPACE_DEPTH`) would probably be better replaced by "ndc"
* I would have created a sub-directory under `Rendering/` and move all IBL shadows related files to this directory
* Try to avoid GC in methods called each frame (as the `update()` method)
* All shader code should probably be in `Shaders/` and not directly in .ts files (cf. )
* Perhaps we should find an abbreviation other than "sss" for "screen space shadows", because it conflicts with "sub-surface scattering".
* I don't think the current code support dynamic update of the scene (adding/removing meshes) after the ibl shadow renderer has been enabled?

As for your question about implementation, I think a rendering pipeline would indeed be better suited, as it would better isolate the code from the rest of the engine.
We're going to try and move away from the existing architecture with scene components, so it's best not to rely on them...

I also wonder about performance...
With the default values of 256 for the voxel resolution, and assuming 8 draw buffers are supported, it means `(256/8)*3=96` rendering to regenerate the voxelized scene.
Even with a resolution of 64, it's still 24 passes to generate the voxelized scene.
Have you been able to run tests on different scenes of varying complexity and assess the impact on performance?

**Comment by MiiBond** [comment]
Excellent, feedback, @Popov72! Thank you.

Regarding performance, yes, it's a lot of draw calls to generate the voxel grid and it isn't really suited to dynamic scenes for that reason (at least in it's current form). There are optimizations yet to be made but, to be honest, most of the work I have ahead of me is for improving the quality rather than the performance.

Enjoy your time off!

**Comment by MiiBond** [comment]
I turned this into a pipeline but I'm not yet using the `addEffect` functionality and adding the pipeline to the `scene.postProcessRenderPipelineManager`.
I think most of the per-camera stuff that I have can be converted to this system (many are ProceduralTextures right now) but I'm unsure of my AccumulationPass. Right now, it relies on a couple of buffers from the previous frame, including the previous result of the accumulation pass. Is there a way to store off RT's to then be used for the next frame for each camera? It doesn't look like the PostProcessRenderPipeline was designed with this in mind.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts
Comment (RaananW) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts:27]:
```suggestion
import type { Camera } from "core/Cameras/camera";
```

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (RaananW) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts:5]: Sorry, this is on me! Should be ignored, but our post-build script doesn't detect it as a comment. I will fix the script for future builds.
```suggestion
// import { Logger } from "../../Misc/logger";
```

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts
Comment (RaananW) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts:10]: Same thing here
```suggestion
// import { Logger } from "../../Misc/logger";
```

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsSpatialBlurPass.ts
Comment (RaananW) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsSpatialBlurPass.ts:5]:
```suggestion
// import { Logger } from "../../Misc/logger";
```

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelTracingPass.ts
Comment (RaananW) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelTracingPass.ts:5]:
```suggestion
// import { Logger } from "../../Misc/logger";
```

**Review by RaananW — COMMENTED** [review]
Just a few suggestions to get the build running correctly. One task is one me (to fix the post-build script from processing commented code).

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (RaananW) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts:5]: You can ignore these comments (of the commented imports) if you merge from master

File: packages/dev/core/src/Materials/Node/Blocks/Fragment/prePassOutputBlock.ts
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/Blocks/Fragment/prePassOutputBlock.ts:121]: This happens after you did that PR but now this file needs update to also support WebGPU (the changes should be straight forward)

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (deltakosh) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts:17]: maybe flag it as _ then? and also as @internal

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts
Comment (deltakosh) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts:26]: same here

File: packages/dev/core/src/Rendering/prePassRenderer.ts
Comment (deltakosh) [inline packages/dev/core/src/Rendering/prePassRenderer.ts:158]: seems to be a breaking change

File: packages/dev/core/src/Shaders/ShadersInclude/prePassDeclaration.fx
Comment (deltakosh) [inline packages/dev/core/src/Shaders/ShadersInclude/prePassDeclaration.fx:5]: we need to make sure that their webgpu version are also implemented

**Review by deltakosh — DISMISSED** [review]
Pretty impressive! I have no formal feedback besides the Webgpu/nme ones.

This is really cool s..t!

I also agree with @Popov72 on the folder ask (to make sure we keep things well separated),

Additionally a PG to test would be awesome!

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (sebavan) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts:241]: I can see only 1 PP being awaited for readyness but 3 have been created in the file. Same goes for dispose, is it intended?

**Review by sebavan — COMMENTED** [review]
I do not see any kind of big architectural issues anymore !!! This looks great :-)

Agree that the main points now are NME/WebGPU as well as the folder.

This is an amazing progress. Something great for the final review would be to provide a tiny summary of the technique if any are available cause there are so many moving parts it would be hard to know where to look in maintenance ?

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts]: Thanks for pointing this out. I've fixed these now.

File: packages/dev/core/src/Rendering/prePassRenderer.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/prePassRenderer.ts]: This was a mistake. I set a bunch of textures to RGBA so that they render in Spector.js.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts]: Done

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts]: Done

File: packages/dev/core/src/Materials/Node/Blocks/Fragment/prePassOutputBlock.ts
Comment (MiiBond) [reply packages/dev/core/src/Materials/Node/Blocks/Fragment/prePassOutputBlock.ts]: Done

File: packages/dev/core/src/Shaders/ShadersInclude/prePassDeclaration.fx
Comment (MiiBond) [reply packages/dev/core/src/Shaders/ShadersInclude/prePassDeclaration.fx:5]: Done

**Comment by MiiBond** [comment]
I've made some updates over the last few days and I think I've addressed all the PR comments.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelRenderer.ts
Comment (MiiBond) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelRenderer.ts:276]: @sebavan , here I'm trying to get the `pass` shader to copy my mip (`mipTarget`) into the mip chain of the voxel grid. When viewing with Spector.js, it appears that `textureSampler` is set to a completely different texture.

Comment (sebavan) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelRenderer.ts:276]: yup answered on the forum, I ll review again once the fix is in. ping me on the thread once ok and the build is good.

**Comment by MiiBond** [comment]
I'm currently having this error print out every frame:
`GL_INVALID_OPERATION: Two textures of different types use the same sampler location`
It seems to be related to the voxel texture (a sampler3D) being used in the `_IblShadowsVoxelTracingPass`. I don't see any obvious issues with it and it seems to be working but it would be nice to clear that up.
I'm also experiencing an issue with the shadows not rendering in the visualization tests. Maybe related?

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15106/merge/testResults/webgpuplaywright/index.html

**Comment by MiiBond** [comment]
I think I've cleaned up the biggest bugs and this is ready to merge when you guys are okay with it.

**Comment by sebavan** [comment]
Amazing progress !!! I ll reset @deltakosh status no worries. I have planned some time tomorrow to review in more detail. Sorry for the delay.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts
Comment (sebavan) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts:20]: Please use the full path

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts
Comment (sebavan) [inline packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts:29]: this should be exported as it is used in the ctor of the main class ?

Why not using only interface as it is used as a property bag only ?

**Review by sebavan — CHANGES_REQUESTED** [review]
Looks really good from a feature and overall design !!!!

Couple global remarks:
1. tsdoc is missing on a few exported public classes/proerties/functions A pass on all would be great before final approval.
2. In the shader code, are the multiline defines fully supported in our preprocessors ?
3. It looks like WebGPU support is added but not everywhere ?
4. should all shaders be moved in dynamic imports like @deltakosh did in the rest of Babylon ?

Point 3 and 4 could be in subsequent PR if needed.

File: packages/dev/core/src/Shaders/geometry.fragment.fx
Comment (noname0310) [inline packages/dev/core/src/Shaders/geometry.fragment.fx:100]: The wgsl geometry.fragment should also be updated

File: packages/dev/core/src/Materials/Textures/Procedurals/proceduralTexture.ts
Comment (MiiBond) [reply packages/dev/core/src/Materials/Textures/Procedurals/proceduralTexture.ts]: Thank you.
Yes, the check for `_generateMipMaps` was a hack to prevent the mips from being automatically generated. I needed `generateMipMaps` to be set for the texture to behave as if it had mips but I create the mips myself and copy them in later so having them automatically generated by WebGL is a waste.
Maybe there's a better way of doing this?

File: packages/dev/core/src/Materials/Textures/Procedurals/proceduralTexture.ts
Comment (MiiBond) [reply packages/dev/core/src/Materials/Textures/Procedurals/proceduralTexture.ts]: Thank you. I missed that.

File: packages/dev/core/src/Materials/Textures/Procedurals/customProceduralTexture.ts
Comment (MiiBond) [reply packages/dev/core/src/Materials/Textures/Procedurals/customProceduralTexture.ts]: Me neither. That was an unintended change. I'm not even using CustomProceduralTexture...
Thanks.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsAccumulationPass.ts]: Done

File: packages/dev/core/src/ShadersWGSL/pbr.vertex.fx
Comment (MiiBond) [reply packages/dev/core/src/ShadersWGSL/pbr.vertex.fx]: I think so but this wasn't a conscious decision. Auto-formatting did this.
I've reverted it now as I think it makes it harder to read anyway.

File: packages/dev/core/src/ShadersWGSL/default.vertex.fx
Comment (MiiBond) [reply packages/dev/core/src/ShadersWGSL/default.vertex.fx]: Should be, yes. But I've reverted anyway.

File: packages/dev/core/src/Shaders/voxelGrid3dDebug.fragment.fx
Comment (MiiBond) [reply packages/dev/core/src/Shaders/voxelGrid3dDebug.fragment.fx]: Removed.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsImportanceSamplingRenderer.ts]: Good catch. Thank you.

File: packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts
Comment (MiiBond) [reply packages/dev/core/src/Rendering/IBLShadows/iblShadowsRenderPipeline.ts]: Done.

File: packages/dev/core/src/Shaders/geometry.fragment.fx
Comment (MiiBond) [reply packages/dev/core/src/Shaders/geometry.fragment.fx:100]: Done.

**Comment by MiiBond** [comment]
> 1. tsdoc is missing on a few exported public classes/proerties/functions A pass on all would be great before final approval.
> 2. In the shader code, are the multiline defines fully supported in our preprocessors ?
> 3. It looks like WebGPU support is added but not everywhere ?
> 4. should all shaders be moved in dynamic imports like @deltakosh did in the rest of Babylon ?

Both 1 and 2 are done.
For us, WebGPU isn't a high priority yet but it will be in the future so I'd like to skip this for now. When I do work on it, we should be able to speed up voxelization significantly using compute shaders. That'll be awesome as this step is currently too slow to support animation.
I can work on dynamic imports now but I don't think it should hold up merging this PR.

**Comment by sebavan** [comment]
yup nw @deltakosh  can help with this part when he ll be back :-) and in the meantime, let s wait for a green build to merge.

**Comment by MiiBond** [comment]
It's green.

**Review by sebavan — APPROVED** [review]

**Comment by sebavan** [comment]
I will merge on Monday morning :-)

**Pull request merged**
